src/test/memory/test_mmap.rs

```diff
@@ -0,0 +1,434 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+use test_utils::set;
+use test_utils::TestEnvironment as TestEnv;
+
+const MAPLEN: usize = 16;
+
+fn validate_shadow_access(buf: &[u8]) -> Result<(), String> {
+    let mut template = "test_mmapXXXXXX".to_string();
+
+    /* Get a file that we can mmap and write into. */
+    let temp_fd = unsafe { libc::mkstemp(template.as_mut_ptr() as *mut libc::c_char) };
+    test_utils::assert_true_and_errno(temp_fd != -1);
+
+    let fp = unsafe { libc::fdopen(temp_fd, "w+".as_ptr() as *const libc::c_char) };
+
+    /* Make sure there is enough space to write after the mmap. */
+    {
+        let rv = unsafe { libc::posix_fallocate(temp_fd as libc::c_int, 0, buf.len() as i64) };
+        assert_eq!(rv, 0, "{}", test_utils::get_errno_message(rv));
+    }
+
+    // Write buffer contents to file.
+    let mut ret = unsafe { libc::fwrite(buf.as_ptr() as *const libc::c_void, 1, buf.len(), fp) };
+    test_utils::assert_true_and_errno(ret == buf.len());
+
+    // Seek to beginning of file.
+    unsafe { libc::fseek(fp, 0, libc::SEEK_SET) };
+    test_utils::assert_errno_nonneg();
+
+    // Read contents from file.
+    let mut mapbuf = vec![0_u8; buf.len()];
+    ret = unsafe {
+        libc::fread(
+            mapbuf.as_mut_ptr() as *mut libc::c_void,
+            1,
+            mapbuf.len(),
+            fp,
+        )
+    };
+    test_utils::assert_true_and_errno(ret == buf.len());
+
+    assert_eq!(mapbuf[..], buf[..]);
+
+    unsafe { libc::unlink(template.as_ptr() as *const libc::c_char) };
+
+    Ok(())
+}
+
+// Wrapper for type conversion of offset from int64_t to off_t.
+// FIXME: Rust's libc doesn't have mmap2, only mmap and mmap64.
+//   The only difference is the offset param is off_t and off64_t, respectively
+unsafe fn mmap_using_mmap(
```

Comment:
Since we're not testing `mmap2`, we *could* just get rid of this wrapper and the function-pointer passing.

OTOH I wouldn't be opposed to making this an internal nix-like wrapper that just provides a more idiomatic rust interface for calling `mmap`. I think we want to avoid using nix's own `mmap` wrapper directly, in case the implementation ever changes to be something other than a very thin wrapper.

If we do turn this into an idiomatic rust wrapper, probably better to return a `Result`, e.g. by using `nix::errno::Errno::result`